### ISSUE TITLE
Maximum update depth exceeded when `ThemeProvider` is portalled

### ISSUE THREAD

**Issue by GerardasB** [iTwin/iTwinUI#2082]
### Describe the bug (current behavior)

Using a `ReactDOM.createPortal()` with `ThemeProvider` component results in an `Error: Maximum update depth exceeded.` error.
This is initially caught by our e2e tests while trying to bump a minor version of iTwinUI https://github.com/iTwin/appui/pull/855

I'm not certain if the provided minimal repro sample is exactly what we hit in the AppUI repo as it is very tricky to determine what is going on (i.e. removing one of the parent `ThemeProviders` avoids this issue). 

Additionally, memoizing the `ref` of a `Box` used in an iTwinUI `Root` component seems to solve this issue.

### Expected Behavior

`ThemeProvider` should not produce any run-time issues.

### Link to minimal repro

https://stackblitz.com/edit/github-vk35v4?file=src%2FApp.tsx,src%2Fmain.tsx

### Steps To Reproduce

1. Open minimal repro
2. Open dev-tools console
3. Observe the error `Error: Maximum update depth exceeded.`

**Comment by mayank99** [comment]
This seems to fix the issue on user end (see [sandbox](https://stackblitz.com/edit/github-vk35v4-2cs5n2?file=src%2FApp.tsx)). Investigating more deeply to find a proper fix.

```diff
 export default function App() {
  const [el, setEl] = React.useState<HTMLElement>();

  React.useEffect(() => {
-   setEl(document.getElementById("x"));
  }, []);

  return (
    <ThemeProvider>
      <div
       id="x"
+      ref={setEl}
      />
      {el &&
        ReactDOM.createPortal(
          <ThemeProvider>
            <Button>X</Button>
          </ThemeProvider>,
          el
        )}
    </ThemeProvider>
  );
}
```

**Comment by mayank99** [comment]
@GerardasB @ignas-k This should be fixed in [3.11.2](https://github.com/iTwin/iTwinUI/releases/tag/%40itwin%2Fitwinui-react%403.11.2)!
